### PR TITLE
[dev] Fix regression in CSA for packed THD sequence training on SM100.

### LINKED ISSUE #7340: [REGRESSION][dev] DeepSeek-V4 CSA repeated recompilation on forward for THD packed sequences.
**Describe the regression**
 @NVIDIA/mcore-oncall After #7064, the performance for Deepseek-v4 with THD packing is ~10x slower due to repeated recompilation on the forward hot path on B200.

**To Reproduce**
~With THD packing, `total_k_rows` will change every for every samples and will trigger a host-side Dynamo recompiles.~
Edit: this is not related, please see the linked PR for actual fix.

```python
import time, torch
from megatron.core.transformer.experimental_attention_variant.csa_utils.fused_sparse_attention import (
    build_thd_compact_k_layout, pack_thd_compact_k,
)
cu_q = torch.tensor([0, 512, 1024, 2048], dtype=torch.int32, device="cuda")
cu_k = torch.tensor([0, 128, 256, 512], dtype=torch.int32, device="cuda")
torch.cuda.synchronize()
t0 = time.perf_counter()
for total_k in range(500, 540):
    cu, m = build_thd_compact_k_layout(cu_q, cu_k, total_k, 4)
    pack_thd_compact_k(torch.randn(total_k, 64, device="cuda", dtype=torch.bfloat16), m)
torch.cuda.synchronize()
print(f"{(time.perf_counter()-t0)*1e3:.0f} ms")  # before: ~1200 ms; after remove torch.compile: ~200 ms
```

**Previous performance**
A dummy 2B DS-v4 arch with THD seq len 23040, 8 gradient accumulation step. 

Image

**Stack trace/logs**
If applicable, add the stack trace or logs related to the regression.

**Environment (please complete the following information):**
 - I'm using megatron-bridge but with the latest MCore's dev branch.
 - New Megatron-LM commit ID: #7064 after this refactor.

**Proposed fix**
Removing the torch.compile for `build_thd_compact_k_layout` and `pack_thd_compact_k` partially recover the performance, but not entirely due to other triton kernel recompile for thd pack.

**Additional context**
Agent's note:

@torch.compile on build_thd_compact_k_layout and pack_thd_compact_k was added earlier in 0f9c777f7 (#5992, compact BF16/MXFP8 DSA indexer on SM100+).

Intent: fuse the small torch logic (segment padding, row map, index_select) into one compiled graph when shapes are stable — e.g. CUDA graph capture with fixed buffer sizes.

#7064 amplified the cost by making the fused training forward call the compact indexer path inside every layer’s FusedCSAIndexerLossFunc.forward whenever thd_compact_indexer_available (B200 + bf16).

k_indexer_flat.shape[0] is total_k_rows. With packed sequences, that integer changes microbatch to microbatch. Dynamo treats it as a compile-time specialization → new graph per distinct value → host-side compile storm × num_layers per step.

Before #7064, the same compiled functions existed but ran less often on the hot path (e.g. compact layout mainly tied to CUDA-graph workspace prep in Path B, while index combining still used window_idxs + torch). After #7064, Path B + compact Top-K + sequence-major indices became the default fused training pipeline, so the compile decorators landed on the hottest call site.

### PR DESCRIPTION
**Pull request by lkhphuc**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Fix the regression for DeepSeek-v4 CSA on packed THD sequence. Root cause is repeated recompilation in the forward hot path.

## Issue tracking

Fix #7340 

## Contribution process

~2~ 1 commits to reduce recompilation on the THD packed batch:
- ~drop torch.compile on THD compact indexer layout helper~ (this actually has no effect, I removed it and keep original behavior)
- stablize CuTe and Triton JIT keys for packed THD CSA

Screenshot 2026-09-15 at 14 31 33

- pink is before #7064 (with a simple local patch to fixed repeated compile for backward THD to be precise, similar in spirit to #6848)
- orange is after, with forward-compute time ~3x
- ~green is after the first commit (removing torch compile for the small helpers)~
- blue is after commit stabilizing the jit compile keys and achieve similar performance as before (pink - with my local backward fix)
- Loss and Grad norm are identical as before fix by resuming from checkpoint.

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] If this PR adds or changes a GPU kernel (Triton, `jit_fuser`/`torch.compile`, CUDA extension, TE or external-library dispatch, or a scatter/index accumulation), I have added or updated its bit-exact determinism test and registered it in `tests/unit_tests/determinism/kernels/manifest.py` ([guide](https://github.com/NVIDIA/Megatron-LM/blob/main/docs/developer/determinism/testing.md))
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### CODE REVIEW

**Comment by hxbai** [comment]
Hi @lkhphuc, thanks for your contribution! Generally LGTM.

I didn't encounter this before since we use padding by default through args like here https://github.com/NVIDIA/Megatron-LM/blob/dev/examples/moe_recipes/deepseek_v4_flash/gb200/mxfp8_THD4K_128GPU_TP1PP2EP64.yaml#L213-L215

Could you help with two things?
1. verify whether there is performance drop on the padding scenarios
2. try to merge the test cases in the new files into other ones

Thanks!

**Comment by lkhphuc** [comment]
Hi @hxbai thanks for your prompt response. I have moved the tests to 2 existing files, hope that's ok.

For the padding case, I confirm the number reported above was already with padding included.
I use Bridge with Energon-native packing and always padded to max seq length via this config: https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/8ec5c9cdc2bf1af3247a2dbe73696534d6055814/src/megatron/bridge/data/builders/energon.py#L179
I believe this is equivalent to mcore's `  pad_packed_seq_alignment: max`


In the bridge's vlm_steps, we don't actually use this setting `  thd_max_packed_sequences: 8`, which set a static length of `num_docs + 1` for `cu_seqlens` I believe. 
So yeah this + low seqlen of 4096 (effective low number of doc per pack) might be the reason you don't see this issue and there should be no perf regression in this case from this PR change.
In our case, the document per pack fluctuate healthily from 5-10 for a max seqlens of 23k.
Screenshot 2026-09-16 at 13 31 36

**Review by hxbai — APPROVED** [review]
LGTM

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/7343/commits/70d90a498b433258c69117553adaf26794930164

**Comment by lkhphuc** [comment]
@hxbai I ran the `tools/autoformat.sh` and keep only the formatted changes in `test_csa_fused_sparse_attention.py` as reported by the CI lint and ignoring the changes for files unrelated to this PR.

The other CI test failed was also due to unrelated doc links missing something.

**Comment by lkhphuc** [comment]
@yaox12 sorry I manually fixed the lint error, which does not show up on my `autoformat.sh` run locally.

**Comment by lkhphuc** [comment]
Apology. Previous test was failing when run on H100. 
In the latest commit i have fixed the test to use hardware-dependent `topk_width` (64 for sm100+, 128 for lower) using the same flashMLA config as the other test. 
The failed test run successfully on A100.
```
============================================== short test summary info ===============================================
PASSED tests/unit_tests/transformer/experimental_attention_variant/test_csa_fused_sparse_attention.py::test_backward_topk_widths_collapse_to_aligned_buckets[None]
PASSED tests/unit_tests/transformer/experimental_attention_variant/test_csa_fused_sparse_attention.py::test_backward_topk_widths_collapse_to_aligned_buckets[topk_length1]
PASSED tests/unit_tests/transformer/experimental_attention_variant/test_csa_fused_sparse_attention.py::test_backward_topk_aligned_width_unchanged
PASSED tests/unit_tests/transformer/experimental_attention_variant/test_csa_fused_sparse_attention.py::test_backward_topk_padding_values[None]
PASSED tests/unit_tests/transformer/experimental_attention_variant/test_csa_fused_sparse_attention.py::test_backward_topk_padding_values[topk_length1]
=================================== 5 passed, 126 deselected, 23 warnings in 0.66s ===================================
```

**Comment by Connor-XY** [comment]
@lkhphuc, thanks for the hardware-dependent test adjustment and A100 results. The current head 5c4507d still has DCO marked Action Required; please address the details in that check. After that, @hxbai could recheck the test updates, since the existing approval is on 70d90a4 and predates these commits.

**Comment by lkhphuc** [comment]
ah done thank you. (only my agent remember to commit with sign-off :))

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/7343/commits/87c8a248530f1362b833d9c096d2f43cf9f93846

**Comment by lkhphuc** [comment]
:( I fixed lint for the new test.
